gcsfs/extended_gcsfs.py

```diff
@@ -0,0 +1,228 @@
+import logging
+from enum import Enum
+
+from fsspec import asyn
+from google.api_core import exceptions as api_exceptions
+from google.api_core import gapic_v1
+from google.api_core.client_info import ClientInfo
+from google.cloud import storage_control_v2
+from google.cloud.storage._experimental.asyncio.async_grpc_client import AsyncGrpcClient
+
+from gcsfs import __version__ as version
+from gcsfs import zb_hns_utils
+from gcsfs.core import GCSFile, GCSFileSystem
+from gcsfs.zonal_file import ZonalFile
+
+logger = logging.getLogger("gcsfs")
+
+USER_AGENT = "python-gcsfs"
+
+
+class BucketType(Enum):
+    ZONAL_HIERARCHICAL = "ZONAL_HIERARCHICAL"
+    HIERARCHICAL = "HIERARCHICAL"
+    NON_HIERARCHICAL = "NON_HIERARCHICAL"
+    UNKNOWN = "UNKNOWN"
+
+
+gcs_file_types = {
+    BucketType.ZONAL_HIERARCHICAL: ZonalFile,
+    BucketType.NON_HIERARCHICAL: GCSFile,
+    BucketType.HIERARCHICAL: GCSFile,
+    BucketType.UNKNOWN: GCSFile,
+}
+
+
+class ExtendedGcsFileSystem(GCSFileSystem):
+    """
+    This class will be used when GCSFS_EXPERIMENTAL_ZB_HNS_SUPPORT env variable is set to true.
+    ExtendedGcsFileSystem is a subclass of GCSFileSystem that adds new logic for bucket types
+    including zonal and hierarchical. For buckets without special properties, it forwards requests
+    to the parent class GCSFileSystem for default processing.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.grpc_client = None
+        self.storage_control_client = None
+        # initializing grpc and storage control client for Hierarchical and
+        # zonal bucket operations
+        self.grpc_client = asyn.sync(self.loop, self._create_grpc_client)
+        self._storage_control_client = asyn.sync(
+            self.loop, self._create_control_plane_client
+        )
```

Comment:
These creations could be deferred, and certainly the two of then done in a single call to sync() rather than two.